Guide - How to move data to and from cloud providers (#10133)
PrefectHQ/prefect: docs/guides/moving-data.md

=== "Azure"

    [prefect-azure](https://prefecthq.github.io/prefect-azure/) provides blocks for interacting with Azure services.

    ```bash
     pip install -U prefect-azure
    ```

=== "GCP"

    [prefect-gcp](https://prefecthq.github.io/prefect-gcp/) provides blocks for interacting with GCP services.
    
    ```bash
     pip install -U prefect-gcp
    ```

## Register the block types

Register the new block types with Prefect Cloud (or with your self-hosted Prefect server instance):

=== "AWS"

    ```bash
    prefect block register -m prefect_aws  
    ```

=== "Azure"

    ```bash
    prefect block register -m prefect_azure 
    ```

=== "GCP"

    ```bash
    prefect block register -m prefect_gcp
    ```

We should see a message in the CLI that several block types were registered.
If we check the UI, we should see the new block types listed.

## Create a storage bucket

Create a storage bucket in the cloud provider account.
Ensure the bucket is publicly accessible or create a user or service account with the appropriate permissions to fetch and write data to the bucket.

## Create a credentials block

If the bucket is private, there are several options to authenticate:

> 1. Use a worker that is authenticated at deployment run time.

Reviewer: this is a little more subtle - the worker doesn't need to be authenticated, it's the deployment's runtime that needs the authentication (and just authenticating the worker doesn't guarantee the deployment's runtime will inherit that)

Author: Thank you. Updated.

If saving credential details in a block we can use a credentials block specific to the cloud provider or use a more generic secret block.
We can create [blocks](/concepts/blocks/) via the UI or Python code.
Below we'll use Python code to create a credentials block for our cloud provider.

!!! Warning "Credentials safety"

    Reminder, don't store credential values in public locations such as public git platform repositories. 
    In the examples below we use environment variables to store credential values.

=== "AWS"

    ```python
    import os
    from prefect_aws import AwsCredentials

    my_aws_creds = AwsCredentials(
        aws_access_key_id="123abc",
        aws_secret_access_key=os.environ.get("MY_AWS_SECRET_ACCESS_KEY"),
    )
    my_aws_creds.save(name="my-aws-creds-block", overwrite=True)
    ```

=== "Azure"

    ```python
    import os
    from prefect_azure import AzureBlobStorageCredentials

    my_azure_creds = AzureBlobStorageCredentials(
        connection_string=os.environ.get("MY_AZURE_CONNECTION_STRING"),
    )
    my_azure_creds.save(name="my-azure-creds-block", overwrite=True)
    ```

=== "GCP"

>     We recommend using specifying the service account key file contents as a string, rather than the path to the file, because that file will not be available in most production environments.

Reviewer: ```suggestion
    We recommend specifying the service account key file contents as a string, rather than the path to the file, because that file might not be available in your production environments.
```

    ```python
    import os
    from prefect_gcp import GCPCredentials

    my_gcp_creds = GCPCredentials(
        service_account_info=os.environ.get("GCP_SERVICE_ACCOUNT_KEY_FILE_CONTENTS"), 
    )
    my_gcp_creds.save(name="my-gcp-creds-block", overwrite=True)
    ```

Run the code to create the block. We should see a message that the block was created.

## Create a storage block

> Let's create a block for cloud provider using Python code or the UI.

Reviewer: ```suggestion
Let's create a block for the chosen cloud provider using Python code or the UI.
```

In this example we'll use Python code.

=== "AWS"

>     Note that the S3Bucket block is not the same as the S3 block that ships with Prefect. 

Reviewer: should the references to `S3Bucket` be code formatted? I'm inclined to say yes since it's a reference to a code object; either way, a good stylistic point to get agreement on

Author: Agree for all references to code objects. I should either use code formatting or change to S3 Bucket.

>     The S3Bucket block we use in this example is part of the prefect-aws library and provides additional functionality. 

Reviewer: same question for the package reference

        s3_block = S3Bucket.load("my-aws-bucket")
        s3_block.get_directory(
            from_path="my_path_to/my_file.parquet", 
            local_path="my_path_to/my_file.parquet"
        )

    if __name__ == "__main__":
        download_from_s3()
    ```

=== "Azure"

    ```python
    from prefect import flow
    from prefect_azure import AzureBlobStorageCredentials
    from prefect_azure.blob_storage import blob_storage_download

    @flow
    def download_from_azure():
        """Flow function to download data"""
        blob_storage_credentials = AzureBlobStorageCredentials.load(
            name="my-azure-creds-block"
        )
        blob_storage_download(
            blob="my_path_to/my_file.parquet",
            container="my_container",
            blob_storage_credentials=blob_storage_credentials,
        )

    if __name__ == "__main__":
        download_from_azure()
    ```

=== "GCP"

    ```python
    from prefect import flow
    from prefect_gcp.cloud_storage import GcsBucket

    gcs_block = GcsBucket.load("zoom-gcs")
    gcs_block.get_directory(
        from_path="my_path_to/my_file.parquet", 
        local_path="my_path_to/my_file.parquet"

>     )

Reviewer: it seems you forgot to wrap this one in a flow with a main block
